File: src/main/java/space/rogi27/homabric/config/HomabricConfig.java
package space.rogi27.homabric.config;

import me.lortseam.completeconfig.api.ConfigContainer;
import me.lortseam.completeconfig.api.ConfigEntry;
import me.lortseam.completeconfig.api.ConfigGroup;
import me.lortseam.completeconfig.data.Config;
import space.rogi27.homabric.Homabric;
import space.rogi27.homabric.utils.HomePermissionObject;
import space.rogi27.homabric.utils.PlayerObject;

import java.util.HashMap;
import java.util.Map;

public class HomabricConfig extends Config implements ConfigContainer {
    public HomabricConfig() {
        super("homabric");
    }

    @Transitive
    public static class Config implements ConfigGroup {
        @ConfigEntry(comment = "Do not touch this value, it allows mod to \ncheck if config file is outdated or not.")
        private static int configVersion = 2;
        @ConfigEntry(comment = "This option enables alternative command variants like /sethome, /removehome and etc.")
        private static boolean enableOldschoolCommands = true;
        @ConfigEntry(comment = "Sets the maximum amount of homes per player.")
        private static int homesLimit = 2;
        @ConfigEntry(comment = "You can define permissions that will override home limit for the players if they have them." +
                "\nPermission names are transformed to permissions like 'homabric.homelimit.<permissionName>'" +
                "\nExample permission: vip: { max-homes=6 }")
        private static Map<String, HomePermissionObject> permissionsHomeLimit = new HashMap<>();

        @ConfigEntry(comment = "List of players with their homes.")
        private static Map<String, PlayerObject> players = new HashMap<>();

        public static int getConfigVersion() {
            return configVersion;
        }

        public static boolean isOldschoolCommandsEnabled() {
            return enableOldschoolCommands;
        }

        public static int homesLimit() {
            return homesLimit;
        }

        public static Map<String, HomePermissionObject> getPermissionsHomeLimit() {
            return permissionsHomeLimit;
        }

        public static PlayerObject getPlayer(String name) {
            return players.get(name);
        }

        public static Map<String, PlayerObject> getPlayers() {
            return players;
        }

        public static PlayerObject getOrCreatePlayer(String name) {
            if(players.get(name) == null) {
                Homabric.LOGGER.warn("There is no data for " + name + ", creating new entry.");
                players.put(name, new PlayerObject().withData(Map.of()));
                Homabric.reloadConfig();
            }
            return players.get(name);
        }
    }
}

File: src/main/kotlin/space/rogi27/homabric/config/HomabricConfig.kt
package space.rogi27.homabric.config

import org.spongepowered.configurate.objectmapping.ConfigSerializable
import org.spongepowered.configurate.objectmapping.meta.Comment
import org.spongepowered.configurate.objectmapping.meta.Setting
import space.rogi27.homabric.objects.HomePermissionObject

object HomabricConfig : ConfigFile<HomabricConfig.Config>("config.conf", Config::class.java, ::Config) {

    @ConfigSerializable
    class Config {
        @Setting("config-version")
        @Comment(
            "Examples and help: https://github.com/rogi27/Homabric/blob/master/README.md#configuration\n" +
                    "Do not touch this value, it allows mod to check if config file is outdated or not."
        )
        var configVersion: Int = 3

        @Setting("enable-classic-commands")
        @Comment("This option enables alternative command variants like /sethome, /removehome and etc.")
        var classicCommandsEnabled: Boolean = true

        @Setting("home-command-aliases")
        @Comment("Allows you to specify aliases for home teleportation commands.")
        var commandHomeAliases: Array<String> = arrayOf("home", "h")

        @Setting("homes-command-aliases")
        @Comment("Allows you to specify aliases for home management commands.")
        var commandHomesAliases: Array<String> = arrayOf("homes", "hs")

        @Setting("homes-limit")
        @Comment("Sets the maximum amount of homes per player.")
        var homesLimit: Int = 2

        @Setting("teleport-cooldown")
        @Comment(
            "Sets the timeout in seconds before player will be teleported home.\n" +
                    "You can disable this feature by setting it to 0."
        )
        var teleportCooldown: Int = 3

        @Setting("permissions-home-limit")
        @Comment(
            "You can define permissions that will override home limit for the players if they have them.\n" +
                    "Permission names are transformed to permissions like 'homabric.homelimit.<permissionName>'\n" +
                    "Example permission: vip: { max-homes=6 }\n" +
                    "If you want to disable limit use 'homabric.limit.bypass' permission."
        )
        var permissionsHomeLimit: Map<String, HomePermissionObject> = emptyMap()
    }
}